### PR TITLE
Fix tpu service fanout

### PR DESCRIPTION
**Pull request by sakridge**
#### Problem

If `slots_per_epoch` is smaller than `MAX_FANOUT_SLOTS * 2`, then the tpu service can return an error.

#### Summary of Changes

Limit fanout to slots_per_epoch

Fixes #

### CODE REVIEW

File: client/src/tpu_client.rs [hunk 1]
```diff
@@ -187,7 +187,9 @@ impl LeaderTpuCache {
     }
 
     fn fetch_slot_leaders(rpc_client: &RpcClient, start_slot: Slot) -> Result<Vec<Pubkey>> {
-        Ok(rpc_client.get_slot_leaders(start_slot, 2 * MAX_FANOUT_SLOTS)?)
+        let epoch_info = rpc_client.get_epoch_info()?;
+        let fanout = (2 * MAX_FANOUT_SLOTS).min(epoch_info.slots_in_epoch * 2);
```
Comment (jstarry) [inline client/src/tpu_client.rs:191]:
```suggestion
        let fanout = (2 * MAX_FANOUT_SLOTS).min(epoch_info.slots_in_epoch);
```
I don't think we want to multiple by 2 here right? You could be towards the end of an epoch and I think this fails if fetching slot leaders for current epoch + 2?

Comment (sakridge) [reply client/src/tpu_client.rs:191]: yea. good point

**Review by jstarry — DISMISSED** [review]
Thanks! Sorry I never tested on a new cluster

File: client/src/tpu_client.rs [hunk 2]
```diff
@@ -187,7 +187,9 @@ impl LeaderTpuCache {
     }
 
     fn fetch_slot_leaders(rpc_client: &RpcClient, start_slot: Slot) -> Result<Vec<Pubkey>> {
-        Ok(rpc_client.get_slot_leaders(start_slot, 2 * MAX_FANOUT_SLOTS)?)
+        let epoch_info = rpc_client.get_epoch_info()?;
```
Comment (mvines) [inline client/src/tpu_client.rs:190]: Let's cache `slots_in_epoch` please to reduce RPC server load, we don't need to be continually asking RPC for this value

Comment (sakridge) [reply client/src/tpu_client.rs]: ok, now it should just call once per epoch.

Comment (mvines) [reply client/src/tpu_client.rs]: Thanks, that's much better than what I was thinking of :)

**Review by mvines — APPROVED** [review]

**Pull request merged**